src/aspire/source/__init__.py

```diff
@@ -114,51 +116,49 @@ def states(self):
 
     @states.setter
     def states(self, values):
-        return self.set_metadata('_rlnClassNumber', values)
+        self.set_metadata('_rlnClassNumber', values)
 
     @property
-    def filters(self):
-        return self.get_metadata('__filter')
-
-    @filters.setter
-    def filters(self, values):
-        self.set_metadata('__filter', values)
-        if values is None:
-            new_values = np.nan
+    def filter_indices(self):
+        return self.get_metadata('__filter_indices')
+
+    @filter_indices.setter
+    def filter_indices(self, indices):
+        # create metadata of filters for all images
+        if indices is None:
+            filter_values = np.nan
         else:
-            new_values = np.array([(
-                getattr(f, 'voltage', np.nan),
-                getattr(f, 'defocus_u', np.nan),
-                getattr(f, 'defocus_v', np.nan),
-                getattr(f, 'defocus_ang', np.nan),
-                getattr(f, 'Cs', np.nan),
-                getattr(f, 'alpha', np.nan)
-            ) for f in values])
+            filter_values = np.array([(
```

Comment:
instead of repeatedly indexing into `self.unique_filters` with `i` can you just do it once? You may have to write this as a for loop instead of a list comprehension

Reply:
Done in the new commit e3393d7.